src/txacme/test/test_challenges.py

```diff
@@ -172,4 +184,115 @@ def test_contains(self, underlay, overlay, key):
+        self.assertThat(
+            zone.list_records(),
+            MatchesListwise([
+                MatchesStructure(
+                    name=EndsWith(subdomain),
+                    type=Equals('TXT'),
+                    )]))
+
+        # Starting twice before stopping doesn't break things
+        self.assertThat(
+            responder.start_responding(server_name, challenge, response),
+            succeeded(Always()))
+        self.assertThat(zone.list_records(), HasLength(1))
+
+        self.assertThat(
+            responder.stop_responding(server_name, challenge, response),
+            succeeded(Always()))
+        self.assertThat(zone.list_records(), HasLength(0))
+
+    @example(token=b'BWYcfxzmOha7-7LoxziqPZIUr99BCz3BfbN9kzSFnrU',
+             subdomain=u'acme-testing',
+             zone_name=u'example.com')
+    @given(token=s.binary(min_size=32, max_size=32).map(b64encode),
+           subdomain=ts.dns_names(),
+           zone_name=ts.dns_names())
+    def test_wrong_zone(self, token, subdomain, zone_name):
+        """
+        Trying to respond for a domain not in the configured zone results in a
+        `.NotInZone` exception.
+        """
+        challenge = self._challenge_factory(token=token)
+        response = challenge.response(RSA_KEY_512)
+        responder = self._responder_factory(zone_name=zone_name)
+        server_name = u'{}.{}.junk'.format(subdomain, zone_name)
+        self.assertThat(
+            maybeDeferred(
+                responder.start_responding, server_name, challenge, response),
+            failed_with(MatchesAll(
+                IsInstance(NotInZone),
+                MatchesStructure(
+                    server_name=EndsWith(server_name),
+                    zone_name=Equals(zone_name)))))
+
+    @example(token=b'BWYcfxzmOha7-7LoxziqPZIUr99BCz3BfbN9kzSFnrU',
+             subdomain=u'acme-testing',
+             zone_name=u'example.com')
+    @given(token=s.binary(min_size=32, max_size=32).map(b64encode),
+           subdomain=ts.dns_names(),
+           zone_name=ts.dns_names())
+    def test_missing_zone(self, token, subdomain, zone_name):
+        """
+        `.ZoneNotFound` is raised if the configured zone cannot be found at the
+        configured provider.
+        """
+        challenge = self._challenge_factory(token=token)
+        response = challenge.response(RSA_KEY_512)
+        responder = self._responder_factory(zone_name=zone_name)
+        server_name = u'{}.{}'.format(subdomain, zone_name)
+        for zone in responder._driver.list_zones():
+            zone.delete()
+        self.assertThat(
+            maybeDeferred(
+                responder.start_responding, server_name, challenge, response),
+            failed_with(MatchesAll(
+                IsInstance(ZoneNotFound),
+                MatchesStructure(
+                    zone_name=Equals(zone_name)))))
+
+
+__all__ = ['TLSResponderTests', 'MergingProxyTests', 'LibcloudResponderTests']
```

Comment:
I'm generating API docs for the test suite, it's mainly for the benefit of that. Possibly useless, but nothing will ever render the docstrings if I don't.